src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -55,71 +60,221 @@ protected override string CreateRequestId()
+                return false;
+            }
+
+            var pathEncoded = pathSegment.IndexOf('%') >= 0;
+
+            // Borrowed from Http1Connection.OnOriginFormTarget
+
+            // URIs are always encoded/escaped to ASCII https://tools.ietf.org/html/rfc3986#page-11
+            // Multibyte Internationalized Resource Identifiers (IRIs) are first converted to utf8;
+            // then encoded/escaped to ASCII  https://www.ietf.org/rfc/rfc3987.txt "Mapping of IRIs to URIs"
+            string requestUrlPath = null;
+
+            try
+            {
+                // The decoder operates only on raw bytes
+                var pathBuffer = new byte[pathSegment.Length].AsSpan();
+                for (int i = 0; i < pathSegment.Length; i++)
+                {
+                    var ch = pathSegment[i];
+                    // The header parser should already be checking this
+                    Debug.Assert(20 < ch && ch < 127);
+                    pathBuffer[i] = (byte)ch;
+                }
+
+                if (pathEncoded)
+                {
+                    // URI was encoded, unescape and then parse as UTF-8
+
+                    var pathLength = UrlDecoder.DecodeInPlace(pathBuffer);
+
+                    // Removing dot segments must be done after unescaping. From RFC 3986:
+                    //
+                    // URI producing applications should percent-encode data octets that
+                    // correspond to characters in the reserved set unless these characters
+                    // are specifically allowed by the URI scheme to represent data in that
+                    // component.  If a reserved character is found in a URI component and
+                    // no delimiting role is known for that character, then it must be
+                    // interpreted as representing the data octet corresponding to that
+                    // character's encoding in US-ASCII.
+                    //
+                    // https://tools.ietf.org/html/rfc3986#section-2.2
+                    pathLength = PathNormalizer.RemoveDotSegments(pathBuffer.Slice(0, pathLength));
+
+                    requestUrlPath = GetUtf8String(pathBuffer.Slice(0, pathLength));
+                }
+                else
+                {
+                    var pathLength = PathNormalizer.RemoveDotSegments(pathBuffer);
+
+                    if (pathSegment.Length == pathLength)
+                    {
+                        // If no decoding was required, no dot segments were removed and
+                        // there is no query, the request path is the same as the raw target
+                        requestUrlPath = pathSegment.ToString();
+                    }
+                    else
+                    {
+                        requestUrlPath = pathBuffer.Slice(0, pathLength).GetAsciiStringNonNullCharacters();
+                    }
+                }
+            }
+            catch (InvalidOperationException)
```

Comment:
It would be nice if we created a shared TryParseTarget method and didn't rely on catching an exception for invalid targets if possible.

Reply:
Unifying the code here is reasonable. That exception only comes from the decoders, I'm not going to try re-writing them at the moment.

Reply:
I could see this being not worth the effort to rewrite the entire DecodePath method not to throw InvalidOperationException internally. I still think it's clearer to move the `catch (InvalidOperationException)` inside the DecodePath method. Just rename it to TryDecodePath and return false in the the catch.

Reply:
In the Http1 code path it's not just DecodePath inside the try/catch, it's also RawTarget and Query ASCII decoding

Reply:
https://github.com/aspnet/KestrelHttpServer/issues/2737